### PR TITLE
Fix version caching

### PR DESCRIPTION
**Pull request by dylant-da**
`daml install latest` does not retrieve a list of versions to install from because of the default `update-check: never` in `$HOME/.daml/daml-config.yaml`, which interacts incorrectly with version retrieval in `daml-assistant/src/DA/Daml/Assistant/Version.hs` and `daml-assistant/src/DA/Daml/Assistant/Cache.hs`, which was added in https://github.com/digital-asset/daml/pull/17029

We change this now:
- When `daml install latest` or `daml version --force-reload` is invoked, it overrides the caching behaviour specified in `.daml/daml-config.yaml`. This preserves the original behaviour before seen.
- When any function requests a list of versions, the list is at least retrieved once, regardless of caching in `.daml/daml-config.yaml`
- Clarify which functions should and shouldn't fail when a list is not available from Github

### CODE REVIEW

File: daml-assistant/src/DA/Daml/Assistant/Cache.hs [hunk 1]
```diff
@@ -63,21 +63,25 @@ cacheAvailableSdkVersions
...
+                -- When UpdateCheckNever, still refresh the cache if it doesn't exist
+                UpdateCheckNever -> Nothing
+                UpdateCheckEvery timeout -> Just timeout
+          | otherwise = Just (CacheTimeout 86400)
+    case mbTimeout of
+      Nothing -> pure ([], Stale)
+      Just timeout ->
+        cacheWith cachePath versionsKey timeout
+            serializeVersions deserializeVersions
+            getVersions
```
Comment (samuel-williams-da) [inline daml-assistant/src/DA/Daml/Assistant/Cache.hs:84]: This function doesn't call loadFromCache anymore - is that a problem?

File: daml-assistant/src/DA/Daml/Assistant/Version.hs [hunk 2]
```diff
@@ -268,25 +272,26 @@ instance FromJSON ParsedSdkVersion where
...
-    versionE <- tryAssistant (getAvailableReleaseVersions useCache)
-    pure $ do
-      (versions, _cacheAge) <- eitherToMaybe versionE
-      maximumMay versions
+freshMaximumOfVersions :: IO ([SdkVersion], CacheAge) -> IO (Maybe SdkVersion)
+freshMaximumOfVersions getVersions = do
+    (versions, cacheAge) <- getVersions
+    case cacheAge of
+      Stale -> pure Nothing
+      Fresh -> pure (maximumMay versions)
```
Comment (samuel-williams-da) [inline daml-assistant/src/DA/Daml/Assistant/Version.hs:288]: You could lift that pure out

**Review by samuel-williams-da — APPROVED** [review]
Have you verified this fixes the problem?

File: daml-assistant/src/DA/Daml/Assistant/Cache.hs (hunk 1 above)
Comment (akrmn) [reply daml-assistant/src/DA/Daml/Assistant/Cache.hs:84]: I was also surprised by this but then saw that `cacheWith` does call `loadFromCache` as its first action, so I think this is actually cleaner

Comment (dylant-da) [reply daml-assistant/src/DA/Daml/Assistant/Cache.hs:84]: Yeah - I decided to move the logic into cacheWith - this code in general had a few spots where logic used by a positive branch was being reimplemented by a negative branch.

**Comment by dylant-da** [comment]
> Have you verified this fixes the problem?

Locally, yes - could someone try it out on their machine as well, to make sure I'm not mad?

The process
```sh
# Reset the daml home dir
> curl -sSL https://get.daml.com/ | sh -s 2.7.1
# Verify that only 2.7.1 lives in ~/.daml
> ls ~/.daml
# Verify that installing latest version fails
> daml install latest
daml: Failed to get latest release version from github.com
  context: Installing the SDK.
> cd daml/ # change directory for whatever repository you cloned this branch to
> daml-sdk-head --skip-jars # install daml-head
# Verify that daml-head can install the latest version
> daml-head install latest
```

I would like to have a test for this behaviour - I'm looking at reverse engineering some of the integration tests to see if we can use that.

This also *does not* work for `daml install latest --snapshots yes`, nor does it work for `daml version --snapshots yes --all yes`.

**Comment by dylant-da** [comment]
Sam has tested that it works on his machine. Making two tickets:
- Solve snapshots issue
- Write compatibility test in `compatibility/`
- Potentially put version list on get.daml.com

**Pull request merged**